prov/sm2/src/sm2_ep.c

```diff
@@ -0,0 +1,1024 @@
+	srx->peer_srx.peer_ops = &sm2_srx_peer_ops;
+
+	srx->peer_srx.ep_fid.fid.fclass = FI_CLASS_SRX_CTX;
+	srx->peer_srx.ep_fid.fid.context = srx;
+	srx->peer_srx.ep_fid.fid.ops = &sm2_srx_fid_ops;
+	srx->peer_srx.ep_fid.ops = &sm2_srx_ops;
+
+	srx->peer_srx.ep_fid.msg = &sm2_srx_msg_ops;
+	srx->peer_srx.ep_fid.tagged = &sm2_srx_tag_ops;
+	*rx_ep = &srx->peer_srx.ep_fid;
+
+	return FI_SUCCESS;
+
+err:
+	free(srx);
+	return ret;
+}
+
+int sm2_srx_context(struct fid_domain *domain, struct fi_rx_attr *attr,
+		    struct fid_ep **rx_ep, void *context)
+{
+	struct sm2_domain *sm2_domain;
+
+	sm2_domain =
+		container_of(domain, struct sm2_domain, util_domain.domain_fid);
+
+	if (attr->op_flags & FI_PEER) {
+		sm2_domain->srx =
+			((struct fi_peer_srx_context *) (context))->srx;
+		sm2_domain->srx->peer_ops = &sm2_srx_peer_ops;
+		return FI_SUCCESS;
+	}
+	return sm2_ep_srx_context(sm2_domain, attr->size, rx_ep);
+}
+
+static int sm2_ep_ctrl(struct fid *fid, int command, void *arg)
+{
+	struct sm2_attr attr;
+	struct sm2_domain *domain;
+	struct sm2_ep *ep;
+	struct sm2_av *av;
+	int ret, self_gid;
+
+	ep = container_of(fid, struct sm2_ep, util_ep.ep_fid.fid);
+	av = container_of(ep->util_ep.av, struct sm2_av, util_av);
+
+	switch (command) {
+	case FI_ENABLE:
+		if ((ofi_needs_rx(ep->util_ep.caps) && !ep->util_ep.rx_cq) ||
+		    (ofi_needs_tx(ep->util_ep.caps) && !ep->util_ep.tx_cq))
+			return -FI_ENOCQ;
+		if (!ep->util_ep.av)
+			return -FI_ENOAV;
+
+		attr.name = ep->name;
+		attr.flags = ep->util_ep.caps & 0;
+
+		ret = sm2_create(&sm2_prov, &attr, &av->mmap, &self_gid);
+		ep->gid = self_gid;
+
+		if (ret)
+			return ret;
+
+		if (!ep->srx) {
+			domain = container_of(ep->util_ep.domain,
+					      struct sm2_domain,
+					      util_domain.domain_fid);
+			// TODO Figure out what size goes here?? ep->tx_size?
+			ret = sm2_ep_srx_context(domain, ep->tx_size, &ep->srx);
+			if (ret)
+				return ret;
+			ret = sm2_srx_bind(&ep->srx->fid,
+					   &ep->util_ep.rx_cq->cq_fid.fid,
+					   FI_RECV);
+			if (ret)
+				return ret;
+		} else {
+			ep->util_ep.ep_fid.msg = &sm2_no_recv_msg_ops;
+			ep->util_ep.ep_fid.tagged = &sm2_no_recv_tag_ops;
+		}
+
+		break;
+	default:
+		return -FI_ENOSYS;
+	}
+	return ret;
+}
+
+static struct fi_ops sm2_ep_fi_ops = {
+	.size = sizeof(struct fi_ops),
+	.close = sm2_ep_close,
+	.bind = sm2_ep_bind,
+	.control = sm2_ep_ctrl,
+	.ops_open = fi_no_ops_open,
+};
+
+static int sm2_endpoint_name(struct sm2_ep *ep, char *name, char *addr,
+			     size_t addrlen)
+{
+	memset(name, 0, SM2_NAME_MAX);
+	if (!addr || addrlen > SM2_NAME_MAX)
+		return -FI_EINVAL;
+
+	pthread_mutex_lock(&sm2_ep_list_lock);
+	ep->ep_idx = sm2_global_ep_idx++;
+	pthread_mutex_unlock(&sm2_ep_list_lock);
+
+	if (strstr(addr, SM2_PREFIX))
+		snprintf(name, SM2_NAME_MAX - 1, "%s:%d:%d", addr, getuid(),
+			 ep->ep_idx);
+	else
+		/* this is an fi_ns:// address.*/
+		snprintf(name, SM2_NAME_MAX - 1, "%s", addr);
```

Comment:
add brackets for comment addition

Reply:
done, if you bracket one, you bracket them all.